src/input-states.js

```diff
@@ -0,0 +1,256 @@
+    switch ( keys.length ) { 
+
+        case 1: { 
+            if (keys[0] === 'inportName' ) {
+                return InputType.IIP;
+            }
+            throw new Error("Invalid input state IIP port information specified!");
+        }
+
+        case 3: { 
+            // Make sure we really have the 3 portInfo keys we want, and not 3 other keys 
+            var lookupKeys = _.pick( portInfo, ALL_PORTINFO_KEYS ); 
+            if ( lookupKeys && _.keys( lookupKeys ).length === 3  ) {
+               return InputType.PACKET;
+            }
+            throw new Error("Invalid input state packet port information specified!");
+        }
+
+        default: {
+            throw new Error("Invalid input state port information found!");
+        }
+    }
+}
+
+module.exports = { 
+
+    /** 
+     * count the number of input states for the specified port.  This will count
+     * any state associated with the portname - both IIP and packet input states.
+     *
+     * @param portName name of port 
+     *
+     * @return the number of input states associated with that port.
+     */
+    count: function( portName ) { 
+        var states = _.where( inputStates, { inportName: portName } );
+        return ( _.isEmpty( states ) ? 0 : states.length );
+    },
+
+    /**
+     * Delete all input port states 
+     */
+    deleteAll: function() { 
+        inputStates.splice( 0, inputStates.length );
+    },
+
+    /**
+     * Delete the state associated with the specified port.
+     *
+     * @param portInfo the port information for the state to retrieve.  All portInfo objects must
+     *                 have an inportName.  Optionally, if the state is from another component and
+     *                 was therefore received in a packet, the portInfo may include: 
+     *                 <ul> 
+     *                    <li> sourceNodeName (optional) name of the node that sent the new state data </li>
+     *                    <li> sourcePortName (optional) name of the the source node port that sent the state data.</li>
+     *                 </ul>
+     * @throws an error if the portInfo is not valid.
+     */
+    delete: function( portInfo ) { 
+
+        verifyPortInfoType( portInfo );
+        var index = getStateIndex( portInfo ); 
+
+        if ( index !== -1 ) { 
+            inputStates.splice(index, 1);
+        }
+ 
+        return this;
+    },
+
+    /** 
+     * Get all input data for the specified port.  If there are multiple inputs on this port,
+     * an array of the input data will be returned.
+     *
+     * @param portName name of port for which all state data should be retrieved
+     *
+     * @return all port data associated with the portName
+     */
+    getData: function( portName ) { 
+
+        var states = _.pluck( _.where( inputStates, { inportName: portName } ), 'state' );
+
+        if ( _.isEmpty( states ) ) { 
+            return {};
+        } 
+
+        if ( states.length === 1 ) { 
```

Comment:
This should check for non-multi instead of states.length .

Somewhere there should also be error checking: (a) to ensure that required ports are attached; and (b) to detect multiple attachments to a non-multi port.   It probably should not be done here, so I've created issue 24 for it: https://github.com/rdf-pipeline/noflo-rdf-pipeline/issues/24